### PR TITLE
🛂 Update Handling of Gem Permissions

### PR DESCRIPTION
**Pull request by landongrindheim**
Until now, we've allowed a `Rubygem` to be owned by users (via `Ownership`s) and an `Organization`. This has some downsides. For example, we've had to explicitly check both ownerships and organization membership in our gem-related policies.

I've opted to push this logic into `Rubygem` and enforce the ownership check to be organization-specific if the gem belongs to an organization OR ownership-specific (i.e. user-specific) if it is not.

### CODE REVIEW

**Comment by pboling** [comment]
A downside of allowing Orgs to own gems is they have a host of distinct issues related to their own state changes that differ from the state changes that occur with people.

Keeping it as just people may be simpler than coming up with policy for how to handle a gem owned by a company that:

* goes out of business
* merges, or is acquired by, another company
* conflicting international business laws and which country's would take precedent
* lawsuits over IP, or publishing rights, that a company claims to "own"

When ownership is limited to people a policy of "for the good of the community and ecosystem" can be more easily determined and adjudicated by the rubygems team.

Orgs may complicate things. Offloading that complexity onto orgs as is currently done, may do less harm than allowing them to own gems.

Due to the additional work involved in managing Org owned gems, it might be a way for RubyGems to generate sponsorships.

Org ownership could be constrained by payment of a monthly fee, which would help alleviate the additional support/maintenance/hosting costs associated with expanding ownership to allow for orgs.

Org ownership of gems could fall back to person ownership upon non-payment of the fee.

IMO, in case it wasn't obvious, for all of the above.

File: app/models/rubygem.rb [hunk 1]
```diff
@@ -186,13 +186,17 @@ def indexed_versions?
...
-    ownerships.exists?(user_id: user.id) || (owned_by_organization? && user_authorized_for_organization?(user))
+    owned_by_with_role?(user, :maintainer)
   end
 
   def owned_by_with_role?(user, minimum_required_role)
     return false if user.blank?
-    ownerships.user_with_minimum_role(user, minimum_required_role).exists?
+
+    if owned_by_organization?
+      organization.memberships.where(user: user).with_minimum_role(minimum_required_role).exists?
```
Comment (colby-swandale) [inline app/models/rubygem.rb]: This will leave out the current implementation of Outside Contributors, which uses the `ownerships` table.

File: app/models/gem_permissions.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    when :owner
+      rubygem.organization.owned_by?(user)
+    else
+      false
+    end
+  end
+
+  def user_in_role?(minimum_role)
+    rubygem.ownerships.user_with_minimum_role(user, minimum_role).exists?
+  end
```
Comment (landongrindheim) [inline app/models/gem_permissions.rb:54]: I think it makes sense to leave these here _for now_, but I foresee a future where org- (and team-) based permissions are checked in a separate class.

**Comment by colby-swandale** [comment]
I feel that this Pull Request needs a bit more time to bake before I'm ready to accept it. This is a good step, but I also feel it's not complete either. With Organisation teams coming down the pipeline I feel it would be a good opportunity to spend a bit more time on this?

**Comment by landongrindheim** [comment]
> I feel that this Pull Request needs a bit more time to bake before I'm ready to accept it. This is a good step, but I also feel it's not complete either. With Organisation teams coming down the pipeline I feel it would be a good opportunity to spend a bit more time on this?

@colby-swandale I think you're right to be cautious with core permission logic. The organization teams work _will_ add complexity here.

Would it be helpful if I walked through how this is designed to extend? I can show how the permission methods would handle the additional cases we'll need, or we can schedule a quick call to discuss the technical approach.

Happy to adjust the design based on our conversations, or hold off if you'd prefer to finalize the broader requirements first 😄

File: app/controllers/api/v1/deletions_controller.rb [hunk 3]
```diff
@@ -45,4 +45,8 @@ def validate_gem_and_version
       end
     end
   end
+
+  def can_delete_gem?(user)
+    GemPermissions.new(@rubygem, user).can_push?
+  end
```
Comment (jenshenny) [inline app/controllers/api/v1/deletions_controller.rb:51]: I think `delete_gem?` would be good to have in `GemPermissions`

File: app/controllers/api/base_controller.rb [hunk 4]
```diff
@@ -94,8 +94,4 @@ def render_bad_request(error = "bad request")
     error = error.message if error.is_a?(Exception)
     render json: { error: error.to_s }, status: :bad_request
   end
-
-  def owner?
```
Comment (jenshenny) [inline app/controllers/api/base_controller.rb]: If this is unused, can we move this to a separate PR since it isn't directly related to the PR?

File: app/models/user.rb [hunk 5]
```diff
@@ -275,8 +275,8 @@ def blocked?
     blocked_email.present?
   end
 
-  def owns_gem?(rubygem)
-    rubygem.owned_by?(self)
+  def can_push?(rubygem)
```
Comment (jenshenny) [inline app/models/user.rb:278]: Do these mean the same? I thought `owns_gem?` would be like an owner while `can_push` is  like a maintainer with lower permissions

File: test/system/invitation_test.rb [hunk 6]
```diff
@@ -25,6 +25,17 @@ class InvitationTest < ApplicationSystemTestCase
     assert_text I18n.t("organizations.members.create.member_invited")
   end
 
+  test "maintainers cannot invite users to an organization" do
```
Comment (jenshenny) [inline test/system/invitation_test.rb:28]: This seems like it can be in another PR too.

File: test/models/gem_permissions_test.rb [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    end
+  end
+
+  context "edge cases" do
+    setup do
+      @rubygem = create(:rubygem)
+      @permissions = GemPermissions.new(@rubygem, @user)
+    end
+
+    should "handle unknown minimum_role gracefully" do
```
Comment (jenshenny) [inline test/models/gem_permissions_test.rb]: It's not clear in this test what the unknown role is here, could you elaborate what this is testing?

**Review by jenshenny — COMMENTED** [review]
I like that the gem permissions logic is more centralized with this PR.

File: app/controllers/api/base_controller.rb (hunk 4 above)
Comment (landongrindheim) [reply app/controllers/api/base_controller.rb]: Yep! I opened https://github.com/rubygems/rubygems.org/pull/6086 for this 😄

File: app/models/user.rb (hunk 5 above)
Comment (landongrindheim) [reply app/models/user.rb:278]: As the code is currently written (i.e. in `main`, but not in this branch), [Rubygem#owned_by?](https://github.com/rubygems/rubygems.org/blob/064ee0b586f3f88d06706e717e651ce5c772f2f7/app/models/rubygem.rb#L188-L191) only considers whether there is an association. It does not consider role, which is why I've chosen to use a name reflecting lower-permissions.

File: test/system/invitation_test.rb (hunk 6 above)
Comment (landongrindheim) [reply test/system/invitation_test.rb:28]: Opended https://github.com/rubygems/rubygems.org/pull/6090 for this 😄

File: test/models/gem_permissions_test.rb (hunk 7 above)
Comment (landongrindheim) [reply test/models/gem_permissions_test.rb]: I don't remember 😳 When I try to use an unknown role, I get the following error:

```
GemPermissionsTest#test_: edge cases should handle unknown minimum_role gracefully. :
RuntimeError: Neutered Exception ActiveRecord::RecordInvalid: Validation failed: Role is not included in the list
```

I'm going to remove this context block as it doesn't add value and I've lost track of intention.

Comment (landongrindheim) [reply test/models/gem_permissions_test.rb]: Removed in [6759a19](https://github.com/rubygems/rubygems.org/pull/5849/commits/6759a191cf588c0bba68253c3ccd7dcd6773d52d)
